frontend/ui/widget/textboxwidget.lua

```diff
@@ -2186,4 +2183,228 @@ function TextBoxWidget:getSourceIndex(char_idx)
+    return word_start_idx, word_end_idx
+end
+
+function TextBoxWidget:getXtextHighlightRects(text_start_idx, text_end_idx, start_line_num, end_line_num)
+    local rects = {}
+
+    for line_num = start_line_num, end_line_num, 1 do
+        local line = self.vertical_string_list[line_num]
+        if line.xglyphs then -- non-empty line
+            local line_last_rect = nil
+
+            for _, xglyph in ipairs(line.xglyphs) do
+                if xglyph.text_index >= text_start_idx and xglyph.text_index <= text_end_idx and not xglyph.no_drawing then
+                    if line_last_rect == nil then
+                        local rect = Geom:new{
+                            x = xglyph.x0,
+                            y = (line_num - self.virtual_line_num) * self.line_height_px,
+                            w = xglyph.x1 - xglyph.x0,
+                            h = self.line_height_px,
+                        }
+
+                        table.insert(rects, rect)
+                        line_last_rect = rect
+                    else
+                        line_last_rect.w = xglyph.x1 - line_last_rect.x
+                    end
+                end
+            end
+        end
+    end
+
+    return rects
+end
+
+function TextBoxWidget:getNonXtextHighlightIndices(start_x, start_y, end_x, end_y)
+    local start_idx = self:_findWordEdge(start_x, start_y, FIND_START)
+    local end_idx = self:_findWordEdge(end_x, end_y, FIND_END)
+    if not start_idx or not end_idx then
+        -- one or both hold points were out of text
+        return nil, nil
+    end
+
+    return start_idx, end_idx
+end
+
+function TextBoxWidget:getNonXtextHighlightRects(text_start_idx, text_end_idx, start_line_num, end_line_num)
+    local rects = {}
+
+    for line_num = start_line_num, end_line_num, 1 do
+        local line = self.vertical_string_list[line_num]
+        if line.end_offset then
+            local line_last_rect = nil
+            local x = 0
+
+            for text_index = line.offset, line.end_offset, 1 do
+                local width = self.char_width[self.charlist[text_index]] + (self.idx_pad[text_index] or 0)
+                if text_index >= text_start_idx and text_index <= text_end_idx then
+                    if line_last_rect == nil then
+                        local rect = Geom:new{
+                            x = x,
+                            y = (line_num - self.virtual_line_num) * self.line_height_px,
+                            w = width,
+                            h = self.line_height_px,
+                        }
+
+                        table.insert(rects, rect)
+                        line_last_rect = rect
+                    else
+                        line_last_rect.w = (x + width) - line_last_rect.x
+                    end
+                end
+
+                x = x + width
+            end
+        end
+    end
+
+    return rects
```

Comment:
(Not a fan of these random blank lines in these 4 functions where it does not really separate bits of work.)

Reply:
I reduced the number of blank lines in the new code.